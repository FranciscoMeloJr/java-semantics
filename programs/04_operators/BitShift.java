class main {

	void printAll(String[] vs, int[] v) {
		for (int i = 0; i<v.length; i++) {
      System.out.println(vs[i]+" = "+v[i]);
		}
	}

  int nrOps = 12;

  String[] initVS() {
    String[] vs = new String[nrOps];
    int i=0;

		vs[i++] = " 4 <<   1    ";
		vs[i++] = " 4 <<  -1    ";
		vs[i++] = "-4 <<   1    ";
		vs[i++] = "-4 <<  -1    ";
		vs[i++] = " 4  >>  1    ";
		vs[i++] = " 4  >> -1    ";
		vs[i++] = "-4  >>  1    ";
		vs[i++] = "-4  >> -1    ";
		vs[i++] = " 4 >>>  1    ";
		vs[i++] = " 4 >>> -1    ";
		vs[i++] = "-4 >>>  1    ";
		vs[i++] = "-4 >>> -1    ";

    return vs;
  }

	static void main(String[] args) {
    int[] v = new int[nrOps];
    String[] vs = initVS();
    int i=0;

		v[i++] =  4 <<   1;
		v[i++] =  4 <<  -1;//different from java due to overflow
		v[i++] = -4 <<   1;
		v[i++] = -4 <<  -1;//different from java due to overflow
		v[i++] =  4  >>  1;
		v[i++] =  4  >> -1;
		v[i++] = -4  >>  1;
		v[i++] = -4  >> -1;
		v[i++] =  4 >>>  1;
		v[i++] =  4 >>> -1;
		v[i++] = -4 >>>  1;
		v[i++] = -4 >>> -1;

    printAll(vs,v);
    System.out.println("Done!");
	}
}
