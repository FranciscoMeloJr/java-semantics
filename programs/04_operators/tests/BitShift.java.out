 4 <<   1     = 8
 4 <<  -1     = 8589934592
-4 <<   1     = -8
-4 <<  -1     = -8589934592
 4  >>  1     = 2
 4  >> -1     = 0
-4  >>  1     = -2
-4  >> -1     = -1
 4 >>>  1     = 2
 4 >>> -1     = 2
-4 >>>  1     = -2
-4 >>> -1     = 1
Done!
